backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -203,6 +203,8 @@ public String getNeededData(EvaluationContext evalContext) {
 
         if (needDatum != null) {
             return needDatum;
+        } else if (entries.isEmpty()) {
+            throw new RuntimeException("Collected datums don't match required datums for entries at command " + currentCmd);
```

Comment:
Originally talked about in https://github.com/dimagi/commcare-core/pull/367#discussion_r74509690 this came up again and was confusing the hell out of me. Failing explicitly instead of with an NPE would probably help avoid such confusion, so opting to add it back